### ISSUE TITLE
plugin for tinc vpn

### ISSUE THREAD

**Issue by startryst** [netdata/netdata#7775]
##### Feature idea summary
As the subject, any plan to do it?

**Comment by thiagoftsm** [comment]
Hi @startryst ,

I never worked with this protocol, but I have experience with other VPNs, so my conclusion is that you want a plugin capable to monitor specific network traffic and number of connections, of course, there are more things we can monitor inside a VPN. I am glad to say that I am working in a collector that will allow users to monitor these traffics that I cited, but it is possible you wanna know more specific things about this VPN, please, could you give us a list of important things that you wanna monitor?

**Comment by startryst** [comment]
so first, let give you the web site of this VPN/tinc, which is pretty popular in the internet, I'll revert back my requirement for that.

And thanks for the reply.

@thiagoftsm

**Comment by Ferroin** [comment]
For detection, check for the presence of `/run/tinc.*.pid`. The `*` part will be the VPN name, and there might be multiple VPN's on any given system.

For actual data collection, either the CLI can be used like `tinc -n <net> <cmd>` with `<net>` replaced with the network name, and `<cmd>` replaced with the specific command, or we could try to use the socket directly (`/run/tinc.*.socket`, `*` is the network name). Both require root access by default.

The most relevant command for data collection is `tinc dump nodes`, which displays a bunch of info about all the nodes in the network, one node per line with info being a simple set of space separated fields like many things in `/proc`. Sample output (with IP's and ID's redacted):

```
aurek id ************* at *.*.*.* port 655 cipher 0 digest 0 maclength 0 compression 0 options 700000c status 08da nexthop aurek via aurek distance 1 pmtu 1439 (min 1439 max 1439) rx 0 0 tx 1 62 rtt 31.225
besh id ************* at *.*.*.* port 655 cipher 0 digest 0 maclength 0 compression 10 options 700000c status 08da nexthop besh via besh distance 1 pmtu 1439 (min 1439 max 1439) rx 654 221257 tx 4205 659680 rtt 34.797
home_one id ************* at *.*.*.* port 655 cipher 0 digest 0 maclength 0 compression 10 options 700000c status 105a nexthop aurek via home_one distance 2 pmtu 1439 (min 0 max 9018) rx 4 1156 tx 8 1438
wild_karde id ************* at MYSELF port 655 cipher 0 digest 0 maclength 0 compression 0 options 700000c status 0058 nexthop wild_karde via wild_karde distance 0 pmtu 9018 (min 0 max 9018) rx 4214 661180 tx 651 221561
```

The most reevant fields are status (not sure what the number means though), rx, tx, and possibly distance (tinc is an ad-hoc mesh topology, so distance may change over time).

----

As of right now, I don't exactly have a bunch of time to work on this myself. If someone else picks it up though, please tag me for review on the PR, as I have interest in seeing this added as well as having the ability to test it since I already use tinc.

**Comment by thiagoftsm** [comment]
I will talk with @ilyam8 to see who can bring this, probably during this sprint.

**Comment by startryst** [comment]
from my perspective, following are the things wish to collect:

1. What's the speed of the tinc tunnel currently running?
since tinc will create the virtual nic, so far the netdata can show the virtual nic's send/receive

2. The stability of the tinc connection.
Tinc will establish a meta-connection(which can think of the control plane), to detect if any problem happens to the tunnel, so that to bring-up/down the tunnel, where so far, I can see those event by tinc log, but if there's any way to show when/how/what reason many times the tunnel/meta-connection is up or down, that'll be super helpful
```
Jan 19 18:05:20 box1 tincd: svr1 (x.x.x.x port 655) didn't respond to PING in 5 seconds
Jan 19 18:05:20 box1 tincd: Closing connection with svr1 (x.x.x.x port 655)
Jan 19 18:05:20 box1 tincd: Could not set up a meta connection to svr1
Jan 19 18:05:20 box1 tincd: Trying to re-establish outgoing connection in 5 seconds
Jan 19 18:05:25 box1 tincd: Trying to connect to svr1 (x.x.x.x port 655)
Jan 19 18:05:25 box1 tincd: Connected to svr1 (x.x.x.x port 655)
Jan 19 18:05:30 box1 tincd: Timeout from svr1 (x.x.x.x port 655) during authentication
Jan 19 18:05:30 box1 tincd: Closing connection with svr1 (x.x.x.x port 655)
Jan 19 18:05:30 box1 tincd: Could not set up a meta connection to svr1
Jan 19 18:05:30 box1 tincd: Trying to re-establish outgoing connection in 10 seconds
Jan 19 18:05:40 box1 tincd: Trying to connect to svr1 (x.x.x.x port 655)
Jan 19 18:05:40 box1 tincd: Connected to svr1 (x.x.x.x port 655)
Jan 19 18:05:45 box1 tincd: Timeout from svr1 (x.x.x.x port 655) during authentication
Jan 19 18:05:45 box1 tincd: Closing connection with svr1 (x.x.x.x port 655)
Jan 19 18:05:45 box1 tincd: Could not set up a meta connection to svr1
Jan 19 18:05:45 box1 tincd: Trying to re-establish outgoing connection in 15 seconds
Jan 19 18:06:00 box1 tincd: Trying to connect to svr1 (x.x.x.x port 655)
Jan 19 18:06:00 box1 tincd: Connected to svr1 (x.x.x.x port 655)
Jan 19 18:06:05 box1 tincd: Timeout from svr1 (x.x.x.x port 655) during authentication
Jan 19 18:06:05 box1 tincd: Closing connection with svr1 (x.x.x.x port 655)
Jan 19 18:06:05 box1 tincd: Could not set up a meta connection to svr1
Jan 19 18:06:05 box1 tincd: Trying to re-establish outgoing connection in 20 seconds
Jan 19 18:06:25 box1 tincd: Trying to connect to svr1 (x.x.x.x port 655)
Jan 19 18:06:25 box1 tincd: Connected to svr1 (x.x.x.x port 655)
Jan 19 18:06:30 box1 tincd: Timeout from svr1 (x.x.x.x port 655) during authentication
Jan 19 18:06:30 box1 tincd: Closing connection with svr1 (x.x.x.x port 655)
Jan 19 18:06:30 box1 tincd: Could not set up a meta connection to svr1
```


3. data plane running TCP or udp?
By default tinc will try to establish tunnel with the other-end by UDP, but if failed, will fall back to TCP, would like to see the graph when the tunnel runs on UDP or TCP.
As for now, I have to tcpdump to verify its running on TCP or UDP

4. Show the tinc routing path
when issue the below command:
`tinned -n netname -kUSR2`
it will generate the output into the tinc log below:
```
Jan 19 21:14:42 box1 tincd: Nodes:
Jan 19 21:14:42 box1 tincd: svrhk1 at x.x.x.x port 443 cipher 91 digest 64 maclength 4 compression 0 options d status 003a nexthop svrhk1 via box1 pmtu 1459 (min 1459 max 1459)
Jan 19 21:14:42 box1 tincd: xxx1 at y.y.y.y port 443 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop svrhk1 via box1 pmtu 1518 (min 0 max 1518)
Jan 19 21:14:42 box1 tincd: xxx9 at z.z.z.z port 443 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop svrhk1 via box1 pmtu 1518 (min 0 max 1518)
Jan 19 21:14:42 box1 tincd: End of nodes.
```
For above, you will see, box1 to srvhk1 is direct, but box1 to xxx1 and xxx9 are routing through svrhk1


I may have other ideas, will post here in the future.
@thiagoftsm  @ilyam8

**Comment by ilyam8** [comment]
@startryst i think the easiest way to start is to check info from all `dump` commands. Could you share? (mangle/remove personal data).

**Comment by startryst** [comment]
> @startryst i think the easiest way to start is to check info from all `dump/info` commands. Could you share? (mangle/remove personal data).

@ilyam8 the dump command you referred is by tinc 1.1 which is a pre-release, the stable one is actually 1.0.x, the documentation for 1.0 is https://www.tinc-vpn.org/documentation/

Regarding the dump info, in v1.0, we usefully use USR2 to dump the tinc info into the syslog, check below for example:

`tincd -n netname -kUSR2`
once we issued above command(where the netname is the VPN name you defined for your  tinc process), then tinc will output the below info into syslog, by default to /var/log/message, but normally we will define something into separate file, like /var/log/tinc.log:
```
Jan 20 20:10:23 box4 tincd: Statistics for Linux tun/tap device (tun mode) /dev/net/tun:
Jan 20 20:10:23 box4 tincd: total bytes in:  21712286834
Jan 20 20:10:23 box4 tincd: total bytes out: 68074057641
Jan 20 20:10:23 box4 tincd: Nodes:
Jan 20 20:10:23 box4 tincd: box2 at 1.1.1.1 port 1112 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
Jan 20 20:10:23 box4 tincd: box3 at 2.2.2.2 port 1113 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
Jan 20 20:10:23 box4 tincd: box4 at MYSELF cipher 0 digest 0 maclength 0 compression 0 options d status 0018 nexthop box4 via box4 pmtu 1518 (min 0 max 1518)
Jan 20 20:10:23 box4 tincd: box5 at 3.3.3.3 port 1125 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
Jan 20 20:10:23 box4 tincd: la2 at 4.4.4.4 port 443 cipher 419 digest 0 maclength 4 compression 0 options d status 003a nexthop la2 via box4 pmtu 1467 (min 1467 max 1467)
Jan 20 20:10:23 box4 tincd: c2gl20 at 5.5.5.5 port 1208 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
Jan 20 20:10:23 box4 tincd: End of nodes.
Jan 20 20:10:23 box4 tincd: Edges:
Jan 20 20:10:23 box4 tincd: box2 to la2 at 4.4.4.4 port 443 options d weight 2585
Jan 20 20:10:23 box4 tincd: box3 to la2 at 4.4.4.4 port 443 options d weight 4891
Jan 20 20:10:23 box4 tincd: box4 to la2 at 4.4.4.4 port 443 options d weight 325
Jan 20 20:10:23 box4 tincd: box5 to la2 at 4.4.4.4 port 443 options d weight 4043
Jan 20 20:10:23 box4 tincd: la2 to box2 at 1.1.1.1 port 1112 options d weight 2585
Jan 20 20:10:23 box4 tincd: la2 to box3 at 2.2.2.2 port 1113 options d weight 4891
Jan 20 20:10:23 box4 tincd: la2 to box4 at 6.6.6.6 port 8012 options d weight 325
Jan 20 20:10:23 box4 tincd: la2 to box5 at 3.3.3.3 port 1125 options d weight 4043
Jan 20 20:10:23 box4 tincd: la2 to c2gl20 at 5.5.5.5 port 1208 options d weight 446
Jan 20 20:10:23 box4 tincd: c2gl20 to la2 at 4.4.4.4 port 443 options d weight 446
Jan 20 20:10:23 box4 tincd: End of edges.
Jan 20 20:10:23 box4 tincd: Subnet list:
Jan 20 20:10:23 box4 tincd: 10.0.12.2/32#10 owner la2
Jan 20 20:10:23 box4 tincd: 10.0.12.104/32#10 owner box4
Jan 20 20:10:23 box4 tincd: 0.0.0.0/0#10 owner la2
Jan 20 20:10:23 box4 tincd: End of subnet list.
```

The earlier input I made in the post(the second point), where you can see the running logs by tinc, shows the authentication error, I would like see is it possible to show the occurrence of those error in the graph?

**Comment by ilyam8** [comment]
> the dump command you referred is by tinc 1.1 which is a pre-release, the stable one is actually 1.0.x

You are right, i just googled `man tinc` and opened first page.

So there is no `dump/info` cli options in the `1.0.x`, it is not good 😞


> shows the authentication error, I would like see is it possible to show the occurrence of those error in the graph?

It is not tinc specific, but take a look

> https://github.com/netdata/netdata/issues/7420

it could do.

___

> we usefully use USR2 to dump the tinc info into the syslog

What do you do next? how do you use the dump?

**Comment by startryst** [comment]
For data plane running TCP or udp?
By default tinc will try to establish tunnel with the other-end by UDP, but if failed, will fall back to TCP, would like to see the graph when the tunnel runs on UDP or TCP.
As for now, I have to tcpdump to verify its running on TCP or UDP

@ilyam8 do we think is there anyway we can explore on this one?

**Comment by ilyam8** [comment]
> As for now, I have to tcpdump to verify its running on TCP or UDP

I see no such info in `/var/log/tinc.log`

I think `netwrok viewer` thing (wip) would be useful, using it you can monitor per proto (tcp/udp) per port traffic (bytes/packets). @thiagoftsm

**Comment by thiagoftsm** [comment]
> I think `netwrok viewer` thing (wip) could be useful. Using it you can monitor per proto (tcp/udp) per port traffic (bytes/packets).

I agree 100%, I am working this week to have per process connection information, so we will have this.

> @thiagoftsm knows better.

**Comment by ilyam8** [comment]
@thiagoftsm per process or per group of processes?

I was asking about something different: connections/bytes/packets per proto per port. If i remember correctly `network viewer` allows to get such metrics.

**Comment by thiagoftsm** [comment]
> @thiagoftsm per process or per group of processes?
> 
> I was asking about something different: connections/bytes/packets per proto per port. If i remember correctly `network viewer` allows to get such metrics.

You are correct, we have this already ready, but I will also do the integration with `apps.plugin` per process. We will have both.

**Comment by startryst** [comment]
Great, look forward to it, I'll check this/next week from the dashboard, which upgrade automatically.

**Comment by startryst** [comment]
Also, another request is to list all different nodes the real-time traffic distribution, like for example tinc server A is use for client B, C, D, what's the B, C, D traffic usage in real-time monitoring...

**Comment by cakrit** [comment]
Looks like this one needs to be planned.

**Comment by ilyam8** [comment]
It doesnt look like ready for planning for me, i am still not clear what is `tinc monitoring`.

> Also, another request is to list all different nodes the real-time traffic distribution, like for example tinc server A is use for client B, C, D, what's the B, C, D traffic usage in real-time monitoring..

`tinc` daemon itself doesnt provide such info, and there is no way to visualize tinc network on the dashboard.

**Comment by startryst** [comment]
> It doesnt look like ready for planning for me, i am still not clear what is `tinc monitoring`.
> 
> > Also, another request is to list all different nodes the real-time traffic distribution, like for example tinc server A is use for client B, C, D, what's the B, C, D traffic usage in real-time monitoring..
> 
> `tinc` daemon itself doesnt provide such info, and there is no way to visualize tinc network on the dashboard.

https://www.tinc-vpn.org/documentation/Signals.html#Signals

`USR2`: Dumps virtual network device statistics, all known nodes, edges and subnets to syslog.

will generate the log I listed above to parse...

**Comment by ilyam8** [comment]
These statistics?

```cmd
Jan 20 20:10:23 box4 tincd: Statistics for Linux tun/tap device (tun mode) /dev/net/tun:
Jan 20 20:10:23 box4 tincd: total bytes in:  21712286834
Jan 20 20:10:23 box4 tincd: total bytes out: 68074057641
Jan 20 20:10:23 box4 tincd: Nodes:
Jan 20 20:10:23 box4 tincd: box2 at 1.1.1.1 port 1112 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
Jan 20 20:10:23 box4 tincd: box3 at 2.2.2.2 port 1113 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
Jan 20 20:10:23 box4 tincd: box4 at MYSELF cipher 0 digest 0 maclength 0 compression 0 options d status 0018 nexthop box4 via box4 pmtu 1518 (min 0 max 1518)
Jan 20 20:10:23 box4 tincd: box5 at 3.3.3.3 port 1125 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
Jan 20 20:10:23 box4 tincd: la2 at 4.4.4.4 port 443 cipher 419 digest 0 maclength 4 compression 0 options d status 003a nexthop la2 via box4 pmtu 1467 (min 1467 max 1467)
Jan 20 20:10:23 box4 tincd: c2gl20 at 5.5.5.5 port 1208 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
Jan 20 20:10:23 box4 tincd: End of nodes.
Jan 20 20:10:23 box4 tincd: Edges:
Jan 20 20:10:23 box4 tincd: box2 to la2 at 4.4.4.4 port 443 options d weight 2585
Jan 20 20:10:23 box4 tincd: box3 to la2 at 4.4.4.4 port 443 options d weight 4891
Jan 20 20:10:23 box4 tincd: box4 to la2 at 4.4.4.4 port 443 options d weight 325
Jan 20 20:10:23 box4 tincd: box5 to la2 at 4.4.4.4 port 443 options d weight 4043
Jan 20 20:10:23 box4 tincd: la2 to box2 at 1.1.1.1 port 1112 options d weight 2585
Jan 20 20:10:23 box4 tincd: la2 to box3 at 2.2.2.2 port 1113 options d weight 4891
Jan 20 20:10:23 box4 tincd: la2 to box4 at 6.6.6.6 port 8012 options d weight 325
Jan 20 20:10:23 box4 tincd: la2 to box5 at 3.3.3.3 port 1125 options d weight 4043
Jan 20 20:10:23 box4 tincd: la2 to c2gl20 at 5.5.5.5 port 1208 options d weight 446
Jan 20 20:10:23 box4 tincd: c2gl20 to la2 at 4.4.4.4 port 443 options d weight 446
Jan 20 20:10:23 box4 tincd: End of edges.
Jan 20 20:10:23 box4 tincd: Subnet list:
Jan 20 20:10:23 box4 tincd: 10.0.12.2/32#10 owner la2
Jan 20 20:10:23 box4 tincd: 10.0.12.104/32#10 owner box4
Jan 20 20:10:23 box4 tincd: 0.0.0.0/0#10 owner la2
Jan 20 20:10:23 box4 tincd: End of subnet list.
```

___

It is possible to chart only following:

```cmd
Jan 20 20:10:23 box4 tincd: total bytes in:  21712286834
Jan 20 20:10:23 box4 tincd: total bytes out: 68074057641
```

There is no way to visualise network/connections (graph).

**Comment by startryst** [comment]
> These statistics?
> 
> ```batchfile
> Jan 20 20:10:23 box4 tincd: Statistics for Linux tun/tap device (tun mode) /dev/net/tun:
> Jan 20 20:10:23 box4 tincd: total bytes in:  21712286834
> Jan 20 20:10:23 box4 tincd: total bytes out: 68074057641
> Jan 20 20:10:23 box4 tincd: Nodes:
> Jan 20 20:10:23 box4 tincd: box2 at 1.1.1.1 port 1112 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
> Jan 20 20:10:23 box4 tincd: box3 at 2.2.2.2 port 1113 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
> Jan 20 20:10:23 box4 tincd: box4 at MYSELF cipher 0 digest 0 maclength 0 compression 0 options d status 0018 nexthop box4 via box4 pmtu 1518 (min 0 max 1518)
> Jan 20 20:10:23 box4 tincd: box5 at 3.3.3.3 port 1125 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
> Jan 20 20:10:23 box4 tincd: la2 at 4.4.4.4 port 443 cipher 419 digest 0 maclength 4 compression 0 options d status 003a nexthop la2 via box4 pmtu 1467 (min 1467 max 1467)
> Jan 20 20:10:23 box4 tincd: c2gl20 at 5.5.5.5 port 1208 cipher 0 digest 0 maclength 0 compression 0 options d status 0038 nexthop la2 via box4 pmtu 1518 (min 0 max 1518)
> Jan 20 20:10:23 box4 tincd: End of nodes.
> Jan 20 20:10:23 box4 tincd: Edges:
> Jan 20 20:10:23 box4 tincd: box2 to la2 at 4.4.4.4 port 443 options d weight 2585
> Jan 20 20:10:23 box4 tincd: box3 to la2 at 4.4.4.4 port 443 options d weight 4891
> Jan 20 20:10:23 box4 tincd: box4 to la2 at 4.4.4.4 port 443 options d weight 325
> Jan 20 20:10:23 box4 tincd: box5 to la2 at 4.4.4.4 port 443 options d weight 4043
> Jan 20 20:10:23 box4 tincd: la2 to box2 at 1.1.1.1 port 1112 options d weight 2585
> Jan 20 20:10:23 box4 tincd: la2 to box3 at 2.2.2.2 port 1113 options d weight 4891
> Jan 20 20:10:23 box4 tincd: la2 to box4 at 6.6.6.6 port 8012 options d weight 325
> Jan 20 20:10:23 box4 tincd: la2 to box5 at 3.3.3.3 port 1125 options d weight 4043
> Jan 20 20:10:23 box4 tincd: la2 to c2gl20 at 5.5.5.5 port 1208 options d weight 446
> Jan 20 20:10:23 box4 tincd: c2gl20 to la2 at 4.4.4.4 port 443 options d weight 446
> Jan 20 20:10:23 box4 tincd: End of edges.
> Jan 20 20:10:23 box4 tincd: Subnet list:
> Jan 20 20:10:23 box4 tincd: 10.0.12.2/32#10 owner la2
> Jan 20 20:10:23 box4 tincd: 10.0.12.104/32#10 owner box4
> Jan 20 20:10:23 box4 tincd: 0.0.0.0/0#10 owner la2
> Jan 20 20:10:23 box4 tincd: End of subnet list.
> ```
> 
> It is possible to chart only following:
> 
> ```batchfile
> Jan 20 20:10:23 box4 tincd: total bytes in:  21712286834
> Jan 20 20:10:23 box4 tincd: total bytes out: 68074057641
> ```
> 
> There is no way to visualise network/connections (graph).

yeah.

**Comment by ilyam8** [comment]
```cmd
Jan 20 20:10:23 box4 tincd: total bytes in:  21712286834
Jan 20 20:10:23 box4 tincd: total bytes out: 68074057641
```

As i said in https://github.com/netdata/netdata/issues/7775#issuecomment-576328725 it can be monitored with `network viewer` thing (when it is available 😄). I think @thiagoftsm is working on getting it to Netdata.

**Comment by thiagoftsm** [comment]
Yes @ilyam8 , until the end of the month, we intend to bring this for Netdata.  ;)

**Comment by Ferroin** [comment]
@ilyam8 If you query the daemon itself, it's possible to get more per-node statistics, including total distance to each known node (which is useful in ad-hoc mesh networks like this). See https://github.com/netdata/netdata/issues/7775#issuecomment-575650094.

**Comment by ilyam8** [comment]
> If you query the daemon itself

Please give some examples. What output, how it can be used. And `dump` command is not available in `1.0` which is current stable

**Comment by startryst** [comment]
> > If you query the daemon itself
> 
> Please give some examples. What output, how it can be used. And `dump` command is not available in `1.0` which is current stable

below is the answer.

https://www.tinc-vpn.org/documentation/Signals.html#Signals

USR2: Dumps virtual network device statistics, all known nodes, edges and subnets to syslog.

**Comment by Ferroin** [comment]
Ah, yeah, I forgot about the whole 1.1 versus 1.0 differentiation.

We obviously need to support 1.0, but there are people using 1.1 out there because it's been in a prerelease state for 8 years now (prerelease 1 for 1.1 is from June of 2011) and is honestly stable as far as most people who don't need government levels of security verification are concerned (AFAICT, lack of formal verification is _why_ it's still listed as prerelease), so it would be nice to have reasonable support for the enhanced statistics that it provides.
